sympy/external/gmpy.py

```diff
@@ -88,80 +89,118 @@
+        return None
+    else:
+        warn(f"Using python-flint {_flint_version} because SYMPY_GROUND_TYPES "
+             f"is set to flint but this version of SymPy has only been tested "
+             f"with python-flint {_PYTHON_FLINT_VERSION_NEEDED}.")
+        return flint
 
-    # Actually import gmpy2
-    gmpy = import_module('gmpy2', min_module_version='2.0.0',
-                module_version_attr='version', module_version_attr_call_args=())
-    flint = None
 
-    if gmpy is None:
-        # Warn if user explicitly asked for gmpy but it isn't available.
-        if GROUND_TYPES != 'auto':
-            from warnings import warn
-            warn("gmpy library is not installed, switching to 'python' ground types")
+def _get_gmpy2(sympy_ground_types):
+    if sympy_ground_types not in ('auto', 'gmpy', 'gmpy2'):
+        return None
 
-        # Fall back to Python if gmpy2 is not available
-        GROUND_TYPES = 'python'
-    else:
-        GROUND_TYPES = 'gmpy'
+    gmpy = import_module('gmpy2', min_module_version=_GMPY2_MIN_VERSION,
+            module_version_attr='version', module_version_attr_call_args=())
 
-elif GROUND_TYPES == 'flint':
+    if sympy_ground_types != 'auto' and gmpy is None:
+        warn("gmpy2 library is not installed, switching to 'python' ground types")
 
-    # Try to use python_flint
-    flint = import_module('flint')
-    gmpy = None
+    return gmpy
 
-    if flint is None:
-        from warnings import warn
-        warn("python_flint is not installed, switching to 'python' ground types")
-        GROUND_TYPES = 'python'
-    else:
-        GROUND_TYPES = 'flint'
 
-elif GROUND_TYPES == 'python':
+#
+# SYMPY_GROUND_TYPES can be flint, gmpy, gmpy2, python or auto (default)
+#
+_SYMPY_GROUND_TYPES = os.environ.get('SYMPY_GROUND_TYPES', 'auto').lower()
+_flint = None
+_gmpy = None
 
-    # The user asked for Python so ignore gmpy2/flint
-    gmpy = None
-    flint = None
-    GROUND_TYPES = 'python'
+#
+# First handle auto-detection of flint/gmpy2. We will prefer flint if available
+# or otherwise gmpy2 if available and then lastly the python types.
+#
+if _SYMPY_GROUND_TYPES in ('auto', 'flint'):
+    _flint = _get_flint(_SYMPY_GROUND_TYPES)
+    if _flint is not None:
+        _SYMPY_GROUND_TYPES = 'flint'
+    else:
+        _SYMPY_GROUND_TYPES = 'auto'
 
-else:
+if _SYMPY_GROUND_TYPES in ('auto', 'gmpy', 'gmpy2'):
+    _gmpy = _get_gmpy2(_SYMPY_GROUND_TYPES)
+    if _gmpy is not None:
+        _SYMPY_GROUND_TYPES = 'gmpy'
+    else:
+        _SYMPY_GROUND_TYPES = 'python'
 
-    # Invalid value for SYMPY_GROUND_TYPES. Warn and default to Python.
-    from warnings import warn
+if _SYMPY_GROUND_TYPES not in ('flint', 'gmpy', 'python'):
     warn("SYMPY_GROUND_TYPES environment variable unrecognised. "
-         "Should be 'python', 'auto', 'gmpy', or 'gmpy2'")
-    gmpy = None
-    flint = None
-    GROUND_TYPES = 'python'
-
+         "Should be 'auto', 'flint', 'gmpy', 'gmpy2' or 'python'.")
```

Comment:
At 167 we might have 'auto' -- was it your intent to raise here?

Reply:
We will not have auto after the block starting at 169:
https://github.com/sympy/sympy/blob/331e683edda8f4d3d9ea4761a99051dfe74d522a/sympy/external/gmpy.py#L162-L179